### PR TITLE
make ThermoRawFileParser run parallel

### PR DESCRIPTION
**Pull request by bernt-matthias**
and 

- add stdio (script returns always exit code 0)
- fix citation

fixes https://github.com/galaxyproteomics/tools-galaxyp/issues/558
should fix https://github.com/galaxyproteomics/tools-galaxyp/issues/557

we could also use the opportunity and update to 1.3.2 ... but I do not have the time to bump the bioconda recipe.

### CODE REVIEW

**Comment by martenson** [comment]
We could postpone merging this after https://github.com/bioconda/bioconda-recipes/pull/26729 - the bioconda build is passing now.

**Comment by bernt-matthias** [comment]
Yep. Lets do this. Can you check here https://github.com/compomics/ThermoRawFileParser/blob/master/README.md if there are new parameters that we should add?

**Comment by martenson** [comment]
Maybe the exception data? 

```diff
--- from_file
+++ ThermoRawFileParser.exe --help
@@ -28,10 +28,17 @@
   -z, --noZlibCompression    Don't use zlib compression for the m/z ratios and
                                intensities. By default zlib compression is
                                enabled.
+  -a, --allDetectors         Extract additional detector data: UV/PDA etc
   -l, --logging=VALUE        Optional logging level: 0 for silent, 1 for
                                verbose.
   -e, --ignoreInstrumentErrors
                              Ignore missing properties by the instrument.
+  -x, --includeExceptionData Include reference and exception data
+  -L, --msLevel=VALUE        Select MS levels (MS1, MS2, etc) included in the
+                               output, should be a comma-separated list of
+                               integers ( 1,2,3 ) and/or intervals ( 1-3 ),
+                               open-end intervals ( 1- ) are allowed
+  -P, --mgfPrecursor         Include precursor scan number in MGF file TITLE
   -u, --s3_url[=VALUE]       Optional property to write directly the data into
                                S3 Storage.
   -k, --s3_accesskeyid[=VALUE]
```

**Comment by martenson** [comment]
1.3.2 is ready https://anaconda.org/bioconda/thermorawfileparser/files

**Comment by bernt-matthias** [comment]
@martenson the last test should fail (parquet generation) .. and succeed with https://github.com/bioconda/bioconda-recipes/pull/26756 (I tested this manually as described here https://planemo.readthedocs.io/en/latest/writing_how_do_i.html#test-tools-against-a-package-or-container-in-a-bioconda-pull-request)

**Comment by bernt-matthias** [comment]
Tests do not run at all since we need release_21.01 (for parquet) which would come with https://github.com/galaxyproteomics/tools-galaxyp/pull/561

But we can discuss if we want such a recent profile or if its preferable to omit parquet for now.

Also I'm not 100% sure about the output format specific options which I implemented in the conditional.

File: tools/ThermoRawFileParser/thermo_converter.xml [hunk 1]
```diff
@@ -1,144 +1,201 @@
...
-            label="Ignore missing instrument properties" help="If false, it stops the conversion if instrument properties are missing" />
-
-        <param name="output_metadata_selector" type="select" label="Output metadata" >
+        <param argument="--input" name="input_file" type="data" format="thermo.raw" label="Thermo RAW file" help="" optional="False"/>
+        <conditional name="format_cond">
+            <param argument="--format" name="output_format" type="select" label="Output format">
+                <option value="0">mgf</option>
+                <option value="1" selected="True">mzml</option>
+                <option value="2">Indexed mzml</option>
+                <option value="3">Parquet</option>
```
Comment (bgruening) [inline tools/ThermoRawFileParser/thermo_converter.xml:50]: Is there a more specific name for it? Parquet is like hdf5, a general concept/format.

Comment (bernt-matthias) [reply tools/ThermoRawFileParser/thermo_converter.xml:50]: Readme says _Apache Parquet: under development_ .. see also https://github.com/compomics/ThermoRawFileParser/issues/19 :) .. so maybe we skip this and choose profile 20.05 or 20.09?

**Comment by martenson** [comment]
xref https://github.com/compomics/ThermoRawFileParser/issues/105

File: tools/ThermoRawFileParser/thermo_converter.xml (hunk 1 above)
Comment (martenson) [reply tools/ThermoRawFileParser/thermo_converter.xml:50]: @bernt-matthias That seems like the better choice imho, wrapping unfinished features is tricky.

**Comment by martenson** [comment]
The `tools/ThermoRawFileParser/test-data/really_small.parquet` file displays as having 0 bytes to me.

File: tools/ThermoRawFileParser/thermo_converter.xml (hunk 1 above)
Comment (bernt-matthias) [reply tools/ThermoRawFileParser/thermo_converter.xml]: I disabled parquet output for now.

**Review by martenson — APPROVED** [review]

File: tools/ThermoRawFileParser/thermo_converter.xml [hunk 2]
```diff
@@ -1,145 +1,202 @@
-<tool id="thermo_raw_file_converter" name="Thermo" version="1.2.3+galaxy0">
+<tool id="thermo_raw_file_converter" name="Thermo" version="@TOOL_VERSION@+galaxy0" profile="21.01">
```
Comment (martenson) [inline tools/ThermoRawFileParser/thermo_converter.xml:1]: Can we drop the profile a bit down then? Is there a required target?

Comment (bgruening) [reply tools/ThermoRawFileParser/thermo_converter.xml:1]: :+1:

File: tools/ThermoRawFileParser/thermo_converter.xml [hunk 3]
```diff
@@ -1,145 +1,202 @@
...
+                <option value="0">mgf</option>
+                <option value="1" selected="True">mzml</option>
+                <option value="2">Indexed mzml</option>
+                <!-- <option value="3">Parquet</option> -->
+            </param>
+            <when value="0">
+                <param argument="--mgfPrecursor" type="boolean" truevalue="--mgfPrecursor" falsevalue="" label="Include precursor scan number in MGF file" help=""/>
+            </when>
+            <when value="1">
+                <param argument="--noZlibCompression" name="zlib_boolean" type="boolean" truevalue="" falsevalue="-z" checked="true"
```
Comment (bgruening) [inline tools/ThermoRawFileParser/thermo_converter.xml:56]: I'm wondering what this implies? Is there a use-case for our users to disable it?

File: tools/ThermoRawFileParser/thermo_converter.xml [hunk 4]
```diff
@@ -1,145 +1,202 @@
...
-        <collection name="output_indexedmzml_collection" type="list" label="${tool.name} on ${on_string}: Indexed mzML collection">
-            <filter>output_format == "2"</filter>
-            <discover_datasets pattern="(?P&lt;designation&gt;.+)\.mzML" ext="mzml" directory="output_folder"/>
-        </collection>
-
-        <collection name="output_metadata_collection" type="list" label="${tool.name} on ${on_string}: metadata collection">
+        <data name="output" format="mzml" label="${tool.name} on ${on_string}">
+            <change_format>
+                <when input="format_cond.output_format" value="0" format="mgf"/>
+                <when input="format_cond.output_format" value="3" format="parquet"/>
```
Comment (bgruening) [inline tools/ThermoRawFileParser/thermo_converter.xml:91]: drop for now?

File: tools/ThermoRawFileParser/thermo_converter.xml (hunk 1 above)
Comment (martenson) [reply tools/ThermoRawFileParser/thermo_converter.xml]: For future re-enabling the format they used seems to be implemented here: https://github.com/compomics/ThermoRawFileParser/blob/c2d4e16347262074c8c998768f306f1ed7043191/Writer/ParquetSpectrumWriter.cs#L134-L153

File: tools/ThermoRawFileParser/thermo_converter.xml (hunk 3 above)
Comment (bernt-matthias) [reply tools/ThermoRawFileParser/thermo_converter.xml:56]: As far as I know different compression schemes can be used for numerical values. Maybe one or the other downstream tool is incompatible with the default ..

File: tools/ThermoRawFileParser/thermo_converter.xml (hunk 4 above)
Comment (bernt-matthias) [reply tools/ThermoRawFileParser/thermo_converter.xml]: doen

File: tools/ThermoRawFileParser/thermo_converter.xml (hunk 2 above)
Comment (bernt-matthias) [reply tools/ThermoRawFileParser/thermo_converter.xml]: > Can we drop the profile a bit down then?

Sure. I went for 20.05

> Is there a required target?

Not really.

**Review by bgruening — APPROVED** [review]

**Comment by martenson** [comment]
Thanks a lot @bernt-matthias! And for the review @bgruening 👍

**Pull request merged**